### ISSUE TITLE
Consolidate documentation

### ISSUE THREAD

**Issue by carlossvg** [ros-realtime/community#41]
ros-realtime.github.io should be the only documentation website for the RTWG docs. Currently, there are different documents in separated github pages. We should consolidate all the documents in a single place so it's easier to find them.

- [ ] Migrate ros-realtime.github.io to mkdocs => @LanderU 
- [ ] Migrate ros2-realtime-examples reports to markdown format => @carlossvg 
- [ ] Migrate to mkdocs https://github.com/ros-realtime/ros2_realtime_benchmarks reports => @LanderU 
- [ ] Migrate https://ros-realtime.github.io/ros2-realtime-examples/ to ros-realtime.github.io => @carlossvg

**Comment by JanStaschulat** [comment]
List of existing documentation: 

Docu on **docs.ros.org**

- https://docs.ros.org/en/rolling/Tutorials/Building-Realtime-rt_preempt-kernel-for-ROS-2.html
- https://docs.ros.org/en/rolling/Tutorials/Demos/Real-Time-Programming.html
- https://docs.ros.org/en/rolling/Tutorials/Miscellaneous/Building-Realtime-rt_preempt-kernel-for-ROS-2.html
- https://docs.ros.org/en/rolling/How-To-Guides/Building-ROS-2-with-Tracing-Instrumentation.html
- https://docs.ros.org/en/rolling/How-To-Guides/Installing-on-Raspberry-Pi.html
- https://docs.ros.org/en/rolling/How-To-Guides/Using-callback-groups.html

- https://docs.ros.org/en/rolling/Concepts/About-Executors.html

Docu on **ros-realtime.github.io**
- https://ros-realtime.github.io/Guides/guides.html
- https://ros-realtime.github.io/Benchmarks/benchmarks.html
- https://ros-realtime.github.io/Subprojects/subprojects.html
- https://ros-realtime.github.io/Related-Projects/related_projects.html
- https://ros-realtime.github.io/Resources/resources.html
- 
Other documentation: 
- ROSCon Real-Time Executor Workshop: https://www.apex.ai/roscon-21

WIP:
- Executors: RT-Priorization: https://github.com/boschresearch/ros2_documentation/blob/feature/how-to-develop-real-time-application/source/How-To-Guides/Developing-a-real-time-application.rst
-

**Comment by JanStaschulat** [comment]
[Document for consolidating real-time documentation](https://docs.google.com/document/d/1FLv5ggb9qODqGIzNT1wHOOR7JrENRBJsFN8XQq78crM/edit?usp=sharing)

**Comment by JanStaschulat** [comment]
Proposal for restructured real-time documentation at docs.ros.org:
The links are pointing to existing pages on docs.ros.org or other places. These documentation pages might need some update.

**Tutorials**
- Beginner
- Intermediate
- Advanced
    - Real-Time
      - [Real-Time Application Configuration](https://github.com/boschresearch/ros2_documentation/blob/feature/how-to-develop-real-time-application/source/How-To-Guides/Developing-a-real-time-application.rst)
      - [Operating System Setup with RT Linux Kernel](https://ros-realtime.github.io/Guides/Real-Time-Operating-System-Setup/rtos_setup.html#)
      - ROS 2 Tracing
- Demos
     - [Pendulum Demo](https://docs.ros.org/en/rolling/Tutorials/Demos/Real-Time-Programming.html)
- Miscellaneous

**How-To-Guides**
- [Building-Realtime-rt_preempt-kernel-for-ROS-2](https://docs.ros.org/en/rolling/Tutorials/Miscellaneous/Building-Realtime-rt_preempt-kernel-for-ROS-2.html)
- [Building-ROS-2-with-Tracing-Instrumentation](https://docs.ros.org/en/rolling/How-To-Guides/Building-ROS-2-with-Tracing-Instrumentation.html)
to be combined with: https://ros-realtime.github.io/Guides/ros2_tracing_trace_and_analyze.html
- [Installing-on-Raspberry-Pi](https://docs.ros.org/en/rolling/How-To-Guides/Installing-on-Raspberry-Pi.html)
- [Using-callback-groups](https://docs.ros.org/en/rolling/How-To-Guides/Using-callback-groups.html)

**Concepts**
- About-Real-Time-In-ROS2
- [Concepts/About-Executors.html](https://docs.ros.org/en/rolling/Concepts/About-Executors.html)

**Comment by JanStaschulat** [comment]
Updated the [working document](https://docs.google.com/document/d/1FLv5ggb9qODqGIzNT1wHOOR7JrENRBJsFN8XQq78crM/edit?usp=sharing) with improved structure.

Please review and add further existing pages in the table in section "Current status". First proposed new structure contains already the links to existing pages (where available), which need to be updated/improved though.

**Comment by christophebedard** [comment]
> * [Building-ROS-2-with-Tracing-Instrumentation](https://docs.ros.org/en/rolling/How-To-Guides/Building-ROS-2-with-Tracing-Instrumentation.html) (merge with: [ros2_trace_and_analyze](https://ros-realtime.github.io/Guides/ros2_tracing_trace_and_analyze.html))

I wouldn't merge these two. Keeping the "Building ROS 2 with Tracing Instrumentation" as a standalone guide allows us to link to _just that_ from anywhere. We could of course mention the other guide at the end as a logical follow-up guide.

**Comment by nightduck** [comment]
> **Tutorials**
> 
> * Beginner
> * Intermediate
> * Advanced
>   * Real-Time
>     * [Real-Time Application Configuration](https://github.com/boschresearch/ros2_documentation/blob/feature/how-to-develop-real-time-application/source/How-To-Guides/Developing-a-real-time-application.rst)
>     * [Operating System Setup with RT Linux Kernel](https://ros-realtime.github.io/Guides/Real-Time-Operating-System-Setup/rtos_setup.html#)
>     * ROS 2 Tracing
> * Demos
>   * [Pendulum Demo](https://docs.ros.org/en/rolling/Tutorials/Demos/Real-Time-Programming.html)

I think Pendulum Demo and Real-time Application configuration could be merged into one (heavily-rewritten) tutorial. The Pendulum could serve as the hello world program for real-time.

**Comment by JanStaschulat** [comment]
Thank you very much for your active discussion in todays RTWG meeting. I updated our [proposal](https://docs.google.com/document/d/1FLv5ggb9qODqGIzNT1wHOOR7JrENRBJsFN8XQq78crM/edit?usp=sharing) for a consolidated real-time documentation. Next step would be that we create and/or update the pages according to our proposal in our [fork of ros2_documentation](https://github.com/ros-realtime/ros2_documentation).

**Comment by JanStaschulat** [comment]
Made an [initial commit](https://github.com/ros-realtime/ros2_documentation/commit/602be3d68009dddafb3e517a5b6c0d747c53dfef) in the branch [update-real-time](https://github.com/ros-realtime/ros2_documentation/tree/feature/update-real-time) in our fork of the ros2_documentation repository. Made a subsection "Real-Time" and added initial pages.

**Comment by JanStaschulat** [comment]
Created draft tutorial: https://github.com/ros-realtime/ros2_documentation/blob/feature/update-real-time/source/Tutorials/Advanced/Real-Time/Real-Time-Processing-with-multiple-threads.rst
Created example minimal_scheduling_real_time_tutorial: https://github.com/ros-realtime/ros2-realtime-examples/pull/10

**Comment by christophebedard** [comment]
@nightduck I think you mentioned during a RTWG meeting a few weeks ago that you were working on porting/moving the "How to use ros2_tracing to trace and analyze an application" tutorial to docs.ros.org. What is the status? I can help.

**Comment by JanStaschulat** [comment]
Hi @christophebedard . We have not created any page yet. The idea is to have a seperate page on "ros2 tracing" outside the Tutorial/Advanced/RealTime tab. So, from my perspective, you can just go ahead. Any help very much appreciated.

**Comment by christophebedard** [comment]
> but we're not submitting a pull request until we've finished other modifications we're making.

The "How to use ros2_tracing to trace and analyze an application" tutorial doesn't really need to wait until everything else has been ported (it's better to make smaller PRs instead of a single big one), but sure, that's fine.

**Comment by JanStaschulat** [comment]
@christophebedard , Yes, I agree. Aas the ros2_tracing tutorial will not be part of the Advanced/Tutorial/Realtime tab, you can go ahead and update the documentation in a separate PR.

**Comment by christophebedard** [comment]
@nightduck do you want to do it, or should I?

**Comment by nightduck** [comment]
I can do it. Will ping you with the new PR later today

**Comment by carlossvg** [comment]
@JanStaschulat I didn't find a PR to add my comments so I will post them here in the meanwhile.

>The real-time setup is tested by counting the number of context switches of other kernel processes.

I would rephrase this. It sounds like we count the context switches for other processes different than the example.

>It should be zero for the real-time callback.

I would clarify it is expected to see some involuntary context switches even with a real-time priority. There might be other threads with higher priority or interrupts preempting the real-time callback execution.

>The complete source file of this tutorial is available [here](https://github.com/ros-realtime/ros2-realtime-examples/tree/rolling/minimal_scheduling/minimal_scheduling_real_time_tutorial.cpp).

The link seems to be broken.

>Then, we assign to the realtime_thread the priority, that was passed by the command line argument. We are not assigning any scheduling parameter to the default_thread, so that the Publisher and the non real-time Subscription will be processed with default Linux settings. 

We should clarify this tutorial is specific for Linux.

>Note, that this construction of the thread function calls the spin()-function of the Executor. That implies that all callbacks that are managed by the Executor will be processed by this thread and regarding the execution timing "inherit" the scheduling policy of this thread. 

I would clarify what we mean here with execution timing. Also, the term inherit could be misleading, it could be confusing when talking about middleware thread which could inherit the parent's thread scheduling attributes. 

We should probably mention what happens to those threads in this example. In this particular thread only the thread we create to spin the executor will have real-time priority while the middleware threads won't.

**Comment by JanStaschulat** [comment]
@carlossvg @christophebedard @nightduck  @razr I created a pull request https://github.com/ros2/ros2_documentation/pull/3756 but put it in DRAFT mode. This shall make it easier for us to comment and update the text.
